test/denormalized_has_many_test.rb

```diff
@@ -141,6 +141,40 @@ def test_deprecated_set_inverse_association_on_cache_hit
     assert_equal item.object_id, associated_record.item.object_id
   end
 
+  def test_returned_records_should_be_readonly_on_cache_hit
+    IdentityCache.with_fetch_read_only_records do
+      Item.fetch(@record.id) # warm cache
+      record_from_cache_hit = Item.fetch(@record.id)
+      assert record_from_cache_hit.fetch_associated_records.all?(&:readonly?)
+    end
+  end
+
+  def test_returned_records_should_be_readonly_on_cache_miss
+    IdentityCache.with_fetch_read_only_records do
+      record_from_cache_miss = Item.fetch(@record.id)
+      assert record_from_cache_miss.fetch_associated_records.all?(&:readonly?)
+    end
+  end
+
+  def test_db_returned_records_should_never_be_readonly
+    IdentityCache.with_fetch_read_only_records do
+      record_from_db = Item.find(@record.id)
+      uncached_records = record_from_db.associated_records
+      refute uncached_records.all?(&:readonly?)
+      record_from_db.fetch_associated_records
+      refute uncached_records.all?(&:readonly?)
```

Comment:
This isn't testing the result of the above line.  Maybe just write it as `assert record_from_db.fetch_associated_records.none?(&:readonly?)` to actually test `fetch_associated_records`.

After that, this test will fail because `fetch_recursively_cached_association` calls `load_and_readonlyify` on the records.